wireless.8
.\" Copyright (c) 2015, 2016, 2017, Gregor Best <gbe@unobtanium.de>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any purpose
.\" with or without fee is hereby granted, provided that the above copyright notice
.\" and this permission notice appear in all copies.
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
.\" REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND
.\" FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
.\" INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
.\" OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
.\" TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF
.\" THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt WIRELESS 8
.Os
.Sh NAME
.Nm wireless
.Nd scan for and configure wireless networks
.Sh SYNOPSIS
.Nm
.Op Ar file
.Sh DESCRIPTION
.Nm
scans for and configures wireless networks on OpenBSD.
It uses
.Xr ifconfig 8
to do most of the heavy lifting and basically just interprets and applies scan results.
.Pp
.Nm
is deprecated since
.Ox 6.4 .
Use the
.Fl join
option of
.Xr ifconfig 8
instead to configure auto-joined wireless networks.
.Sh USAGE
Just run
.Nm
once to scan for known wireless LANs and configure the earliest network in the configuration file, for example:
.Bd -literal -offset indent
$ wireless
\&... wait a bit for scanning and configuration ...
$ dhclient iwn0
.Ed
.Pp
The result of the scan can also be written to a file.
The format of the file is described in
.Sx "CONFIGURATION STATEMENTS"
below.
.Pp
If necessary (see the
.Sx "CONFIGURATION OVERVIEW"
section below),
.Nm
will notify
.Xr wpa_supplicant 8
of the newly established connection by calling
.Pp
.D1 wpa_cli reassoc
.Sh "CONFIGURATION OVERVIEW"
This is an example configuration file:
.Bd -literal -offset indent
device iwn0
dump /var/db/wireless/aps
verbose

open   freifunk.paderborn.net
802.1x eduroam
wpa    "home network" thisismypassword
.Ed
.Pp
There are three kinds of networks,
.Ic open ,
.Ic wpa
and
.Ic 802.1x .
All of these get an SSID as their first parameter.
.Ic wpa
gets the password or hex key as the second parameter.
SSIDs and passwords which contain spaces can be enclosed in single or double quotes.
Regular string quoting rules apply:
.Pp
.D1 wpa \[dq]Hello\e\[dq]foo\[dq] 'I\e'm a password!'
.Pp
You can include files with the
.Ic include
statement:
.Pp
.D1 include "/home/user/.wireless.conf"
.Pp
Macros can be set and used like this:
.Bd -literal -offset indent
mypw = "foobar! 1234#$"
hexkey = 0x092876e79594ae7beda40dd150d17fa215e51f22cb990f9ec640a267c96fbedf
wpa nw1 $mypw
wpa nw2 $hexkey
.Ed
.Pp
The priority of a network depends on the order of networks in the configuration file.
The earlier a network appears, the higher is its priority.
If an SSID is visible multiple times (such as in campus networks), the access point with the strongest RSSI is chosen.
.Sh CONFIGURATION STATEMENTS
This section outlines the available configuration statements.
.Bl -tag -width Ds
.It Ic debug | Ic verbose
Enables debugging or verbose mode.
See
.Sx DIAGNOSTICS
below for details.
.Pp
Both
.Ic debug
and
.Ic verbose
can be specified at the same time.
.It Ic device Ar name
Use
.Ar name
as the device to configure.
.It Ic dump Pa path
Dump a list of configured networks to
.Pa path .
It contains one line for each access point with the following information:
.Pp
.D1 BSSID RSSI encrypted? known? SSID..
.Pp
There is no default path, which means that if no
.Ic dump
statement is given, the list of networks is not written out.
.It Ic include Pa path
Includes the file
.Pa path
at the point where the
.Ic include
statement appears in the file.
.It Ic open Ar nwid
Defines an unencrypted network.
.It Ic wpa Ar nwid Ar password|hexkey
Defines a WPA2 network.
.It Ic 802.1x Ar nwid
Defines a WPA2 network that uses IEEE 802.1x as the key management scheme.
This requires using
.Xr wpa_supplicant 8
to handle the actual key management.
.El
.Sh FILES
.Bl -tag -width "/etc/wireless.conf" -compact
.It Pa "/etc/wireless.conf"
configuration file
.El
.Sh DIAGNOSTICS
If the keyword
.Ic verbose
is given in the configuration file, the name of the network that is being configured will be printed to the standard error stream.
.Pp
The keyword
.Ic debug
enables behaviour that helps debugging.
Its meaning will change without prior notice, so it should not be used outside of development.
.Sh SEE ALSO
.Xr ifconfig 8 ,
.Xr wpa_passphrase 8 ,
.Xr wpa_supplicant 8
.Sh AUTHORS
.Bl -bullet
.It
.An Gregor Best Aq Mt gbe@unobtanium.de
.It
jggimi (pledging)
.It
spedru (improvements to the
.Xr wpa_supplicant 8
interface)
.It
.An -nosplit
.An Klemens Nanni Aq Mt kn@openbsd.org
(improvements to the manpage)
.El


wireless.c
/*
 * Copyright (c) 2015, 2016, 2017, 2018 Gregor Best <gbe@unobtanium.de>
 * Parts Copyright (c) 2013, 2014, OpenBSD project
 *
 * Permission to use, copy, modify, and/or distribute this software for any purpose
 * with or without fee is hereby granted, provided that the above copyright notice
 * and this permission notice appear in all copies.
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
 * REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
 * INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
 * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
 * TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF
 * THIS SOFTWARE.
*/

#include <assert.h>
#include <err.h>
#include <libgen.h>
#include <spawn.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <sys/queue.h>
#include <sys/ioctl.h>
#include <sys/wait.h>

#include <sys/socket.h>
#include <net/if.h>
#include <netinet/in.h>
#include <netinet/if_ether.h>

#include <net80211/ieee80211.h>
#include <net80211/ieee80211_ioctl.h>

#include "config.h"

#define ARRAY_SIZE(x) ((sizeof(x)) / sizeof(*x))
#define SCANSZ 512

/* Selects the first network in `cnf->networks` that is in `nr` by ESSID and
 * annotates it with the bssid of the scanned AP. The selected network is then
 * returned
 */
struct network *
select_network(struct config *cnf, struct ieee80211_nodereq *nr, int numr) {
	int i;

	for (i = 0; i < numr; i++) {
		struct network *n;

		TAILQ_FOREACH(n, &cnf->networks, networks) {
			if (!strncmp(n->nwid, (char*) nr[i].nr_nwid, IEEE80211_NWID_LEN)) {
				memcpy(n->bssid, nr[i].nr_bssid, IEEE80211_ADDR_LEN);
				return n;
			}
		}
	}

	return NULL;
}

/* This function comes from OpenBSD's /usr/src/sbin/ifconfig/ifconfig.c */
int
rssicmp(const void *a, const void *b) {
	const struct ieee80211_nodereq *n1 = a, *n2 = b;
	int r1 = n1->nr_rssi;
	int r2 = n2->nr_rssi;

	return r2 < r1 ? -1 : r2 > r1;
}

#define DPRINTF(...) do { \
	if (cnf->debug) { \
		fprintf(stderr, __VA_ARGS__); \
	} \
} while (0)

int
scan(struct config *cnf, struct ieee80211_nodereq *nr, int nrlen) {
	struct ieee80211_nodereq_all na;
	struct ifreq ifr;
	int s;

	assert(nrlen > 0);

	DPRINTF("device %s up\n", cnf->device);

	if ((s = socket(AF_INET, SOCK_DGRAM, 0)) < 0)
		err(1, "socket");

	memset(&ifr, 0x00, sizeof(ifr));
	(void) strlcpy(ifr.ifr_name, cnf->device, sizeof(ifr.ifr_name));

	DPRINTF("running SIOCS80211SCAN ioctl\n");

	if (ioctl(s, SIOCS80211SCAN, &ifr) != 0)
		err(1, "ioctl");

	DPRINTF("SIOCS80211SCAN done\n");

	memset(&na, 0x00, sizeof(na));
	memset(nr, 0x00, sizeof(nr) * nrlen);
	na.na_node = nr;
	na.na_size = nrlen * sizeof(*nr);
	(void) strlcpy(na.na_ifname, cnf->device, sizeof(na.na_ifname));

	DPRINTF("running SIOCG80211ALLNODES ioctl\n");

	if (ioctl(s, SIOCG80211ALLNODES, &na) != 0)
		err(1, "ioctl");

	DPRINTF("SIOCG80211ALLNODES done\n");

	if (pledge("stdio proc exec rpath wpath cpath", NULL) == -1) {
		perror("pledge");
		exit(2);
	}

	if (close(s) != 0)
		err(1, "close");

	qsort(nr, na.na_nodes, sizeof(*nr), rssicmp);

	return na.na_nodes;
}

void
configure_network(struct config *cnf, struct network *nw) {
	pid_t child;
	struct ether_addr ea;
	char *params[13]; /* Maximum number of ifconfig/wpa_cli parameters */

	if (!nw) {
		return;
	}

	memcpy(&ea.ether_addr_octet, nw->bssid, sizeof(ea.ether_addr_octet));

	/* Common parameters for all configuration options */
	params[0] = "ifconfig";
	params[1] = cnf->device;
	params[2] = "nwid";
	params[3] = nw->nwid;
	params[4] = "bssid";
	params[5] = ether_ntoa(&ea);
	params[6] = "-chan"; /* Autoselect channel */

	/* three options: open wifi, wpa/wpa2 or 802.1X */
	switch (nw->type) {
		case NW_OPEN:
			params[7] = "-wpa";
			params[8] = "-wpakey";
			params[9] = NULL;
			break;
		case NW_WPA2:
		case NW_8021X:
			params[7] = "wpa";
			params[8] = "wpaakms";
			if (nw->type == NW_WPA2) {
				params[9] = "psk";
				params[10] = "wpakey";
				params[11] = nw->wpakey;
				params[12] = NULL;
			} else {
				params[9] = "802.1x";
				params[10] = NULL;
			}
			break;
		default:
			errx(1, "BUG: Unknown network type %d!\n", nw->type);
	}

	if (cnf->verbose) {
		fprintf(stderr, "configuring wireless network %s\n", nw->nwid);
	}

	if (posix_spawn(&child, "/sbin/ifconfig", NULL, NULL, params, NULL) != 0) {
		err(1, "posix_spawn: ifconfig");
	}
	waitpid(child, NULL, 0);

	if (nw->type == NW_8021X) {
		params[0] = "wpa_cli";
		params[1] = "reassoc";
		params[2] = NULL;

		if (posix_spawn(&child, "/usr/local/sbin/wpa_cli", NULL, NULL, params, NULL) != 0) {
			err(1, "posix_spawn: wpa_cli");
		}

		waitpid(child, NULL, 0);
	}
}

void
write_nwlist(struct config *cnf, struct ieee80211_nodereq *nr, int numnodes) {
	int i, fd, ret;
	char *tmpname;

	if (!cnf->dump) {
		DPRINTF("Won't dump network list, dump not configured\n");
		return;
	}

	DPRINTF("Dumping scanned networks to %s now\n", cnf->dump);

	ret = asprintf(&tmpname, "%s/%s.XXXXXXXXXX", dirname(cnf->dump), basename(cnf->dump));
	if ((!tmpname) || (ret == -1)) {
		err(1, NULL);
	}

	DPRINTF("tmpname=%s\n", tmpname);

	/* Write out /tmp/nw-aps */
	fd = mkstemp(tmpname);
	if (fd < 0) {
		err(1, "mkstemp");
	}

	dprintf(fd, "BSSID% 12c\tRSSI\tENC\tKNOWN\tNWID\n", ' ');

	for (i = 0; i < numnodes; i++) {
		char nwid[IEEE80211_NWID_LEN + 1];
		struct ether_addr ea;
		int len, enc, known;
		struct network *n;

		nwid[IEEE80211_NWID_LEN] = 0x00;
		len = nr[i].nr_nwid_len;
		if (len > IEEE80211_NWID_LEN)
			len = IEEE80211_NWID_LEN;
		memcpy(nwid, nr[i].nr_nwid, len);
		nwid[len] = 0x00;

		memcpy(&ea.ether_addr_octet, nr[i].nr_bssid, sizeof(ea.ether_addr_octet));

		enc = nr[i].nr_capinfo & IEEE80211_CAPINFO_PRIVACY;

		known = 0;
		TAILQ_FOREACH(n, &cnf->networks, networks) {
			if (!strncmp(n->nwid, (char*) nr[i].nr_nwid, IEEE80211_NWID_LEN)) {
				known = 1;
				break;
			}
		}

		/* bssid signal strength enc? known? nwid */
		dprintf(fd, "%s\t%d\t%s\t%sknown\t%s\n", ether_ntoa(&ea), nr[i].nr_rssi,
		        enc? "enc": "open", known? "": "un", nwid);
	}

	close(fd);

	if (rename(tmpname, cnf->dump) == -1)
		err(1, "rename");

	free(tmpname);
}

void
free_config(struct config *cnf) {
	struct network *n, *tmp;

	TAILQ_FOREACH_SAFE(n, &cnf->networks, networks, tmp) {
		free(n->nwid);
		free(n->wpakey);
		free(n);
	}

	free(cnf->device);
	free(cnf);
}

int
main(int argc, char **argv) {
	struct config *cnf;
	struct ieee80211_nodereq nr[SCANSZ];
	int numnodes;

	if (argc == 2)
		cnf = parse_config(argv[1]);
	else
		cnf = parse_config("/etc/wireless.conf");

	if (!cnf) {
		return 1;
	}

	if (!cnf->device) {
		free_config(cnf);
		errx(1, "No device specified");
	}

	memset(nr, 0x00, ARRAY_SIZE(nr));
	numnodes = scan(cnf, nr, ARRAY_SIZE(nr));

	configure_network(cnf, select_network(cnf, nr, numnodes));

	write_nwlist(cnf, nr, numnodes);

	free_config(cnf);

	return 0;
}
